test/functional/mempool_spend_coinbase.py

```diff
@@ -20,40 +20,41 @@
 class MempoolSpendCoinbaseTest(BitcoinTestFramework):
     def set_test_params(self):
         self.num_nodes = 1
-        self.setup_clean_chain = True
 
     def run_test(self):
         wallet = MiniWallet(self.nodes[0])
 
-        wallet.generate(200)
-        chain_height = self.nodes[0].getblockcount()
-        assert_equal(chain_height, 200)
+        # Invalidate two blocks, so that miniwallet has access to a coin that will mature in the next block
+        chain_height = 198
+        self.nodes[0].invalidateblock(self.nodes[0].getblockhash(chain_height + 1))
+        assert_equal(chain_height, self.nodes[0].getblockcount())
 
         # Coinbase at height chain_height-100+1 ok in mempool, should
         # get mined. Coinbase at height chain_height-100+2 is
         # too immature to spend.
-        b = [self.nodes[0].getblockhash(n) for n in range(101, 103)]
-        coinbase_txids = [self.nodes[0].getblock(h)['tx'][0] for h in b]
-        utxo_101 = wallet.get_utxo(txid=coinbase_txids[0])
-        utxo_102 = wallet.get_utxo(txid=coinbase_txids[1])
+        wallet.scan_blocks(start=chain_height - 100 + 1, num=1)
+        utxo_mature = wallet.get_utxo()
+        wallet.scan_blocks(start=chain_height - 100 + 2, num=1)
+        utxo_immature = wallet.get_utxo()
```

Comment:
I don't really like this pattern. The method name `sync_blocks` seems confusing because I wouldn't expect a state change (`self._utxos`) by this method. It seems this would have to be used very carefully, with knowledge of `MiniWallet`'s internals to avoid bugs. For example, if `scan_blocks` is called multiple times with any overlapping blocks, then we'll start to get duplicate utxos in `self._utxos` (maybe this should be a set instead of a list)? 

In general I also don't really like how `self._utxos` is sorted (in `send_self_transfer` by value), while here it isn't sorted but will have an implicit ordering based on when/how `scan_blocks` is called, and then based on order of blocks/vouts inside `scan_blocks`. There's just a lot of different ways it's ordered and we are always relying on that ordering

I'm gonna think this through, don't think this should hold up this PR. But at least I can start the discussion and see what you think / if this can be improved.

Reply:
I agree. Do you have specific suggestions to improve that pass with all existing tests? Feel free to ping me for review.